misk-core/src/main/kotlin/misk/backoff/Retries.kt

```diff
@@ -2,10 +2,17 @@ package misk.backoff
 
 /**
  * Retries the provided function up to a certain number of times, applying the given backoff
- * between each retry. The retry function is provided with current retry count, in case this is
- * relevant
+ * between each retry. If provided, the onRetry callback is called when a retry happens, allowing
+ * clients to perform a task (log, emit metrics) every time a retry occurs.
+ * The retry function is provided with current retry count, in case this is relevant.
  */
-fun <A> retry(upTo: Int, withBackoff: Backoff, f: (retryCount: Int) -> A): A {
+@JvmOverloads
+fun <A> retry(
+  upTo: Int,
+  withBackoff: Backoff,
+  onRetry: ((retryCount: Int, exception: Exception) -> Unit)? = null,
+  f: (retryCount: Int) -> A,
```

Comment:
```suggestion
  block: (retryCount: Int) -> A,
```

Let's rename this to block to be more idiomatic, f isn't used as a named parameter anywhere so this won't break anything.

https://github.com/search?q=org%3Asquareup+%22%2C+f+%3D%22+language%3AKotlin+&type=code

Reply:
I agree !